lib/Sema/CSSimplify.cpp

```diff
@@ -5833,6 +5833,60 @@ ConstraintSystem::simplifyApplicableFnConstraint(
     return simplified;
   }
 
+  // Handle applications of types with `callAsFunction` methods.
+  if (desugar2->mayHaveMembers()) {
+    auto &ctx = getASTContext();
+    // Get all `callAsFunction` methods of the nominal type.
+    // Note: Consider caching `callAsFunction` methods.
+    SmallVector<FuncDecl *, 4> callMethods;
+    auto candidates = lookupMember(desugar2, DeclName(ctx.Id_callAsFunction));
+    for (auto entry : candidates) {
+      auto callMethod = dyn_cast<FuncDecl>(entry.getValueDecl());
+      if (!callMethod)
+        continue;
+      callMethods.push_back(callMethod);
+    }
+
+    // Handle `callAsFunction` methods calls.
+    if (!callMethods.empty()) {
+      // Create a type variable for the `callAsFunction` method.
+      auto loc = getConstraintLocator(locator);
+      auto tv = createTypeVariable(loc, TVO_CanBindToLValue);
+
+      // Record the `callAsFunction` method overload set.
+      SmallVector<OverloadChoice, 4> choices;
+      for (auto candidate : callMethods) {
+        TC.validateDecl(candidate);
+        if (candidate->isInvalid()) continue;
+        choices.push_back(
+            OverloadChoice(type2, candidate, FunctionRefKind::SingleApply));
+      }
+      if (choices.empty()) return SolutionKind::Error;
+      addOverloadSet(tv, choices, DC, loc);
+
+      // Create type variables for each parameter type.
+      SmallVector<AnyFunctionType::Param, 4> tvParams;
+      for (unsigned i : range(func1->getNumParams())) {
+        auto param = func1->getParams()[i];
+        auto paramType = param.getPlainType();
+
+        auto *tvParam = createTypeVariable(loc, TVO_CanBindToNoEscape);
+        auto locatorBuilder =
+            locator.withPathElement(LocatorPathElt::getTupleElement(i));
+        addConstraint(ConstraintKind::ArgumentConversion, paramType,
+                      tvParam, locatorBuilder);
+        tvParams.push_back(AnyFunctionType::Param(
+            tvParam, Identifier(), param.getParameterFlags()));
+      }
+      // Create target function type and an applicable function constraint.
+      AnyFunctionType *funcType =
+          FunctionType::get(tvParams, func1->getResult());
+      addConstraint(ConstraintKind::ApplicableFunction, funcType, tv, locator);
+
+      return SolutionKind::Solved;
+    }
```

Comment:
Same thing: please extract this into a helper function; you can have it return whether it succeeded.

Reply:
Done in https://github.com/apple/swift/pull/24299/commits/fd458d32869960744d0dda8390df0d3bc93efa2d.